form/create_custom_choice_type.rst

```diff
@@ -0,0 +1,273 @@
+        {
+            $resolver
+                // Using whatever way you want to get the choices - Mode::getCategories() is just an example
+                ->setDefault('choices', Model::getCategories())
+
+                // ... override more choice options or define new ones
+            ;
+        }
+    }
+
+.. caution::
+
+     The ``getParent()`` method is used instead of ``extends``.
+     This allows the type to inherit from both ``FormType`` and ``ChoiceType``.
+
+Loading Lazily Static Custom Choices
+------------------------------------
+
+Sometimes, the callable to define the ``choices`` option can be a heavy process
+that could be prevented when the submitted data is optional and empty.
+Sometimes it can depend on other options.
+
+The solution is to load the choices lazily using the ``choice_loader`` option,
+which accepts a callback::
+
+    use Symfony\Component\Form\ChoiceList\ChoiceList;
+    use Symfony\Component\OptionsResolver\Options;
+
+    // use this option instead of the "choices" option
+    ->setDefault('choice_loader', ChoiceList::lazy($this, static function() {
+        return Model::getCategories();
+    }))
+
+    // or if it depends on other options
+    ->setDefault('some_option', 'some_default')
+    ->setDefault('choice_loader', static function (Options $options) {
+        $someOption = $options['some_option'];
+
+        return ChoiceList::lazy($this, static function() use ($someOption) {
+            return Model::getCategories($someOption);
+        }, $someOption);
+    }))
+
+.. note::
+
+    We use the ``ChoiceList::lazy()`` method to create a cached
+    :class:`Symfony\\Component\\Form\\ChoiceList\\Loader\\CallbackChoiceLoader`
+    object. The first argument ``$this`` is the type configuring the form, and
+    a third argument ``$vary`` can be used as array to pass any value that
+    makes the loaded choices different.
+
+Creating a Type With Dynamic Choices
+------------------------------------
+
+When loading choices is complex, a callback is not enough and a "real" service
+is needed. Fortunately, the Form component provides a
+:class:`Symfony\\Component\\Form\\ChoiceList\\Loader\\ChoiceLoaderInterface`.
+You can pass any instance to the ``choice_loader`` option to handle things
+any way you need. For example, you could leverage this new power to load
+categories from an HTTP API. The easiest way is to extend the
+:class:`Symfony\\Component\\Form\\ChoiceList\\Loader\\AbstractChoiceLoader`
+class, which already implements the interface and avoids triggering your logic
+when it is not needed (i.e when the form is submitted empty and valid).
+This could look like this::
+
+    // src/Form/ChoiceList/AcmeCategoryLoader.php.
+    namespace App\Form\ChoiceList;
+
+    use App\Api\AcmeApi;
+    use Symfony\Component\Form\ChoiceList\Loader\AbstractChoiceLoader;
+
+    class AcmeCategoryLoader extends AbstractChoiceLoader;
+    {
+        // This must be passed by the type
```

Comment:
```suggestion
        // this must be passed by the type
```

I figured my review already had enough comments to also be very nitpicky about these little changes 🙈

Reply:
Fixed!